File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/Tae Seong Hong/크레인 인형뽑기 게임.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    vector<int> saves;
    int board_row = board.size() - 1;
    for(int move : moves) {
        int col = move - 1;
        
        if(board[board_row][col] == 0)
            continue;
        
        for(int row = 0;row <= board_row;row++) {
            if(board[row][col] != 0) {
                saves.push_back(board[row][col]);
                board[row][col] = 0;
                break;
            }
        }
        
        if(saves.size() > 1) {
            if(saves[saves.size() - 1] == saves[saves.size() - 2]) {
                saves.pop_back();
                saves.pop_back();
                answer += 2;
            }
        }
    }    
    return answer;
    
}

/*
moves = 크레인 이동 위치
board = 보드판

col = moves - 1

*/

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/Hyun-suck Lee/크레인 인형뽑기 게임.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        Game game = new Game(board);
        for (int move : moves) {
            game.moveToBasket(move - 1);
        }
        
        return game.getRemoveCount();
    }
}

class Game {
    private List<Deque<Integer>> board; // col 기준
    private Deque<Integer> basket;
    private int removeCount;
    
    public Game(int[][] board) {
        this.setBoard(board);
        this.basket = new ArrayDeque<>();
        this.removeCount = 0;
    }
    
    public void moveToBasket(int col) {
        Deque<Integer> target = board.get(col);
        if (target.size() == 0) return;
        
        if (target.size() > 0 && basket.peek() == target.peek()) {
            target.pop();
            basket.pop();
            removeCount += 2;
            return;
        }
        
        basket.push(target.pop());
    }
    
    public void setBoard(int[][] arrBoard) {
        this.board = new ArrayList<>();
        
        for (int i = 0; i < arrBoard[0].length; i++) {
            this.board.add(new ArrayDeque<>());
        }
        
        for (int row = arrBoard.length - 1; row >= 0; row--) {
            for (int col = 0; col < arrBoard[0].length; col++) {
                if (arrBoard[row][col] == 0) continue;
                this.board.get(col).push(arrBoard[row][col]);
            }
        }
    }
    
    public int getRemoveCount() {
        return this.removeCount;
    }
}

/*
- 게임맵: N * N 의 공간
- 바구니: 인형 최대 개수
- LIFO 구조임

- 바구니에 같은 모양 겹치면 제거됨

- 인형 없는 곳 잡아도 문제 X

return 겹쳐 제거된 인형 개수
*/